### PR TITLE
Upgrade SLE12SP3 to SLE15 with zVM

### PR DESCRIPTION
**Pull request by mitiao**
Not use 2 chained jobs anymore to perform upgrade on zVM. Using autoyast to install origin system and then perform media upgrade or proxyscc upgrade on it. Upgrade scenarios could be controlled by different autoyast profile of origin system.
- Related ticket: https://progress.opensuse.org/issues/29101
- Needles: https://gitlab.suse.de/openqa/os-autoinst-needles-sles/merge_requests/768
- Verification run:
upgrade by media: http://bartok.arch.suse.de/tests/166 cancelled due to bsc#1084997
upgrade by proxy:  http://bartok.arch.suse.de/tests/168 failed in scc_registration should be other potential issue

### CODE REVIEW

**Comment by coolo** [comment]
There is no guarantee that there is no job run between these 2 jobs - why aren't you doing this in one job changing VERSION in between? Preferably doing the preparation using autoyast

**Comment by mitiao** [comment]
According to @coolo's comment, switching to autoyast preparation

**Comment by mitiao** [comment]
@Soulofdestiny @rwx788 @QingmingSu Could you review this?

File: data/autoyast_sle12/autoyast_sles12sp3_defalut_reg_full.xml [hunk 1]
```diff
@@ -0,0 +1,254 @@
+<?xml version="1.0"?>
+<!DOCTYPE profile>
```
Comment (okurz) [inline data/autoyast_sle12/autoyast_sles12sp3_defalut_reg_full.xml]: Typo in profile name, should be default

File: lib/utils.pm [hunk 2]
```diff
@@ -1026,6 +1029,86 @@ sub exec_and_insert_password {
     send_key "ret";
 }
 
+=head2 shorten_url
+shotren url via schort(s.qa.suse.de)
```
Comment (okurz) [inline lib/utils.pm]: Shorten

File: lib/utils.pm [hunk 3]
```diff
@@ -1026,6 +1029,86 @@ sub exec_and_insert_password {
...
+This mainly use for autoyast url shorten to avoid limit of x3270 xedit
```
Comment (okurz) [inline lib/utils.pm]: This is mainly used

File: tests/autoyast/console.pm [hunk 4]
```diff
@@ -23,14 +23,17 @@ use utils 'power_action';
...
+    # And no need on zVM
+    if (!check_var('BACKEND', 's390x')) {
+        send_key 'ctrl-alt-f2';
+        send_key 'alt-f2';
+        if (!check_screen 'text-login') {
+            record_soft_failure 'bsc#1042554';
+            send_key 'ctrl-alt-delete';
+            power_action('reboot', keepconsole => 0, observe => 0);
+            $self->wait_boot();
+        }
```
Comment (okurz) [inline tests/autoyast/console.pm]: Actually there is another PR open by me to remove this whole workaround section. I think if the fix has reached SLE12SP4 we can simply remove it

**Comment by okurz** [comment]
merged needle MR.

In general thank you very much for being the first to cover an autoyast installation in openQA on s390x as described in https://progress.opensuse.org/issues/11922 .

**Comment by mitiao** [comment]
Thanks for @okurz's review, rebased for fixes and travis check

File: tests/update/patch_before_migration.pm [hunk 5]
```diff
@@ -52,11 +55,15 @@ sub patching_sle {
...
     if (get_var('FULL_UPDATE')) {
         fully_patch_system();
-        type_string "reboot\n";
-        $self->wait_boot(textmode => !is_desktop_installed(), ready_time => 600, bootloader_time => 200);
-
-        # Go back to the initial state, before the patching
-        $self->setup_migration();
+        # Update origin system on zVM that is controlled by autoyast profile and reboot is done by end of autoyast installation
+        # So we skip reboot here after fully patched on zVM to reduce times of reconnection to s390x
+        if (!get_var('UPGRADE_ON_zVM')) {
```
Comment (okurz) [inline tests/update/patch_before_migration.pm]: Please use only uppercase for test variables. I think we do not have mixed lower+upper elsewhere.

Comment (mitiao) [reply tests/update/patch_before_migration.pm]: Fixed.

File: tests/autoyast/installation.pm [hunk 6]
```diff
@@ -132,6 +133,11 @@ sub run {
       = qw(bios-boot nonexisting-package reboot-after-installation linuxrc-install-fail scc-invalid-url warning-pop-up inst-betawarning autoyast-boot);
     push @needles, 'autoyast-confirm'        if get_var('AUTOYAST_CONFIRM');
     push @needles, 'autoyast-postpartscript' if get_var('USRSCR_DIALOG');
+    # Autoyast reboot automatically without confirmation, usually assert 'bios-boot' that is not existing on zVM
+    # So push a needle to check upcoming reboot on zVM that is a way to indicate the stage done
+    push @needles, 'autoyast-stage1-reboot-upcoming' if check_var('BACKEND', 's390x');
```
Comment (rwx788) [inline tests/autoyast/installation.pm]: I still  wonder if we can do anything better than this needle, for example checking serial output. But let's try it out and see how we can improve.

**Comment by Soulofdestiny** [comment]
LGTM

**Pull request merged**